internal/envoy/cluster.go

```diff
@@ -78,14 +79,21 @@ func upstreamValidationSubjectAltName(c *dag.Cluster) string {
 
 func cluster(cluster *dag.Cluster, service *dag.TCPService) *v2.Cluster {
 	c := &v2.Cluster{
-		Name:                 Clustername(cluster),
-		AltStatName:          altStatName(service),
-		ClusterDiscoveryType: ClusterDiscoveryType(v2.Cluster_EDS),
-		EdsClusterConfig:     edsconfig("contour", service),
-		ConnectTimeout:       250 * time.Millisecond,
-		LbPolicy:             lbPolicy(service.LoadBalancerStrategy),
-		CommonLbConfig:       ClusterCommonLBConfig(),
-		HealthChecks:         edshealthcheck(service),
+		Name:           Clustername(cluster),
+		AltStatName:    altStatName(service),
+		ConnectTimeout: 250 * time.Millisecond,
+		LbPolicy:       lbPolicy(service.LoadBalancerStrategy),
+		CommonLbConfig: ClusterCommonLBConfig(),
+		HealthChecks:   edshealthcheck(service),
+	}
+
+	// If routing to service type `ExternalName` then specify upstream here and not via EDS
+	if len(service.ExternalName) > 0 {
+		c.LoadAssignment = ClusterLoadAssignConfig("contour", service)
+		c.ClusterDiscoveryType = ClusterDiscoveryType(v2.Cluster_STRICT_DNS)
```

Comment:
this is a tiny nit, but as this alternation does the same thing, just with different values, i think it would be clearer if the assignments to c.LoadAssignment and c.ClusterDiscoveryType were in the same order.

Maybe consider using a switch, rather than an if, again to make it clearer that this is a selection, not a conditional. eg

```go
switch len(service.ExternalName) {
case 0:
   // external name not set, cluster will be discovered via EDS
   c.ClusterDiscoveryType = ClusterDiscoveryType(v2.Cluster_EDS)
   c.EdsClusterConfig = edsconfig("contour", service)
default:
  // external name set, use hard coded DNS name
  c.ClusterDiscoveryType = ClusterDiscoveryType(v2.Cluster_STRICT_DNS)
  c.LoadAssignment = ClusterLoadAssignConfig("contour", service)
}
```

Reply:
Great point, it's not a this or that scenario, so this is more clear.